### DISCUSSION TITLE
Effort to reduce manual strings for DataStore Names

### DISCUSSION THREAD

**Discussion by mcbride-clint** [amuste/DnetIndexedDb#26] [Ideas]
I'm new to Github contributions so I think I did it right but I have a fork with some changes that I was experimenting with:

[https://github.com/mcbride-clint/DnetIndexedDb/commit/32d5e95d7f333074a8d0c3e31e5151158a939c70](https://github.com/mcbride-clint/DnetIndexedDb/commit/32d5e95d7f333074a8d0c3e31e5151158a939c70)

I added in a few things that shouldn't break previous usage but built upon @Kylar182 suggested code.

 - Created IndexedDbKey and IndexedDbIndex Attributes to decorate classes.  They have properties to fill out optional options.
 - Created another setup Extension method below: 
   - When you add a store of a type it searches that class for the Key and Index Attributes.  Adds them to the store and names the store after the class.
```CSharp
        var indexedDbDatabaseModel = new IndexedDbDatabaseModel()
            .WithName("TestAttributes")
            .WithVersion(1);

        indexedDbDatabaseModel.AddStore<TableFieldDto>();

        options.UseDatabase(indexedDbDatabaseModel);
```

 - Added New overloads for most of the Interop Methods to allow you no longer have to provide a datastore name and it will instead use the given type.    This would reduce the chances of typos and the number of "magic" strings in the code.

  > `GridColumnDataIndexedDb2.DeleteAll<TableFieldDto>();`
  > is equivalent to
  > `GridColumnDataIndexedDb2.DeleteAll("TableFieldDto");`

Curious for any feedback.

**Comment by Kylar182** [comment]
I like the idea, it'd be nice to have the inferred table method and explicit one though in case someone wanted two tables with the same type. Typical scenario for that would be having one as a backup in case the other got pretty full. That's where your other method for getting a max and / or key count of a table comes in. 

Also, let me know when you get this one through and I'd like to piggy back on it and update the examples to show how I use it. I have a Repository Pattern system I use for syncing data between offline first cloud apps.

**Reply by mcbride-clint** [reply]
Yeah having both is good for both the reason you said and for backwards compatibility. 

I'm trying to make a similar offline app and I'm new to indexed db.  Is there any downside to just having two separate databases that have the same inferred data store information instead of setting both stores up independently?

**Reply by Kylar182** [reply]
I think the only downside would be flexibility but that's quite a minimal difference. I've used this library for 5 offline first apps so far. It's been quite helpful.
